params/json.libevm.go

```diff
@@ -39,36 +39,54 @@ type chainConfigWithExportedExtra struct {
 	Extra                      *pseudo.Type `json:"extra"` // `c.extra` is otherwise unexported
 }
 
-// UnmarshalJSON implements the [json.Unmarshaler] interface.
-func (c *ChainConfig) UnmarshalJSON(data []byte) error {
-	switch reg := registeredExtras; {
-	case reg.Registered() && !reg.Get().reuseJSONRoot:
-		return c.unmarshalJSONWithExtra(data)
-
-	case reg.Registered() && reg.Get().reuseJSONRoot: // although the latter is redundant, it's clearer
-		c.extra = reg.Get().newChainConfig()
-		if err := json.Unmarshal(data, c.extra); err != nil {
-			c.extra = nil
-			return err
-		}
-		fallthrough // Important! We've only unmarshalled the extra field.
-	default: // reg == nil
+// UnmarshalJSON implements the [json.Unmarshaler] interface and JSON decodes
+// `data` according to the following:
+//   - extra is not registered:
+//     `data` is decoded into `c` and the extra is ignored.
+//   - extra is registered and the registered reuseJSONRoot field is false:
+//     `data` is decoded into `c` and the "extra" JSON field in `data` is decoded into `c.extra`.
+//   - extra is registered and the registered reuseJSONRoot field is true:
+//     `data` is decoded into `c` and `data` is decoded into `c.extra`.
+func (c *ChainConfig) UnmarshalJSON(data []byte) (err error) {
+	if !registeredExtras.Registered() {
+		// assume there is no extra
 		return json.Unmarshal(data, (*chainConfigWithoutMethods)(c))
 	}
+	extraConstructors := registeredExtras.Get()
+	c.extra = extraConstructors.newChainConfig()
+	reuseJSONRoot := extraConstructors.reuseJSONRoot
+	return UnmarshalChainConfigJSON(data, c, c.extra, reuseJSONRoot)
 }
 
-// unmarshalJSONWithExtra unmarshals JSON under the assumption that the
-// registered [Extras] payload is in the JSON "extra" key. All other
-// unmarshalling is performed as if no [Extras] were registered.
-func (c *ChainConfig) unmarshalJSONWithExtra(data []byte) error {
-	cc := &chainConfigWithExportedExtra{
-		chainConfigWithoutMethods: (*chainConfigWithoutMethods)(c),
-		Extra:                     registeredExtras.Get().newChainConfig(),
-	}
-	if err := json.Unmarshal(data, cc); err != nil {
-		return err
+// UnmarshalChainConfigJSON JSON decodes `data` according to the following.
+//   - `extra` is not nil and `reuseJSONRoot` is false:
+//     `data` is decoded into `config` and the "extra" JSON field in `data` is decoded into `extra`.
+//   - `extra` is not nil and `reuseJSONRoot` is true:
+//     `data` is decoded into `config` and `data` is decoded into `extra`.
+//   - `extra` is nil:
+//     `data` is decoded into `config` and the extra is ignored.
+func UnmarshalChainConfigJSON[T any](data []byte, config *ChainConfig, extra *T, reuseJSONRoot bool) (err error) {
+	err = json.Unmarshal(data, (*chainConfigWithoutMethods)(config))
+	switch {
```

Comment:
Why is a switch needed? Either we are reusing the JSON root or we're not, and each of those may or may not return an error.

See my comment below about double unmarshalling into the `extra` field when not reusing the root, which may have influenced this.

Reply:
The double unmarshaling is fixed, and switch converted to 2 ifs which are indeed perhaps clearer 😉